File: GMCMultiImage/GMCZoomingMultiImageView.m
// GMCZoomingMultiImageView.m
//
// Copyright (c) 2013 Hilton Campbell
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "GMCZoomingMultiImageView.h"
#import "GMCDecompressImageOperation.h"

const CGSize GMCZoomingMultiImageViewPlaceholderSizeDefault = { 55, 55 };

#define fequal(a,b) (fabs((a) - (b)) < FLT_EPSILON)

@interface GMCZoomingMultiImageView () <UIScrollViewDelegate>

@property (nonatomic, strong) UITapGestureRecognizer *doubleTapGestureRecognizer;

@property (nonatomic, strong) UIScrollView *scrollView;
@property (nonatomic, strong) UIImageView *imageView;
@property (nonatomic, strong) UIActivityIndicatorView *loadingIndicatorView;

@property (nonatomic, assign) CGSize previousBoundsSize;

@property (nonatomic, strong) GMCMultiImageRendition *currentRendition;
@property (nonatomic, strong) NSMutableArray *multiImageRenditionFetches;
@property (nonatomic, strong) NSMutableArray *decompressImageOperations;

@property (nonatomic, assign) CGFloat initialZoomScale;
@property (nonatomic, assign) CGFloat otherZoomScale;

@end

@implementation GMCZoomingMultiImageView

- (id)initWithFrame:(CGRect)frame {
    if ((self = [super initWithFrame:frame])) {
        self.scrollView = [[UIScrollView alloc] initWithFrame:self.bounds];
        self.scrollView.scrollsToTop = NO;
        self.scrollView.showsVerticalScrollIndicator = NO;
        self.scrollView.showsHorizontalScrollIndicator = NO;
        self.scrollView.delegate = self;
        self.scrollView.maximumZoomScale = 1;
        self.scrollView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [self addSubview:self.scrollView];
        
        self.imageView = [[UIImageView alloc] init];
        [self.scrollView addSubview:self.imageView];
        
        self.loadingIndicatorView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
        self.loadingIndicatorView.frame = self.bounds;
        self.loadingIndicatorView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [self addSubview:self.loadingIndicatorView];
        
        self.multiImageRenditionFetches = [NSMutableArray array];
        self.decompressImageOperations = [NSMutableArray array];
        
        self.doubleTapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(doubleTap:)];
        self.doubleTapGestureRecognizer.numberOfTapsRequired = 2;
        [self addGestureRecognizer:self.doubleTapGestureRecognizer];
        
        _placeholderSize = GMCZoomingMultiImageViewPlaceholderSizeDefault;
        _scale = [UIScreen mainScreen].scale;
    }
    return self;
}

- (void)setMultiImage:(GMCMultiImage *)multiImage {
    _multiImage = multiImage;
    
    // Reset state
    self.currentRendition = nil;
    self.imageView.image = nil;
    [self.loadingIndicatorView stopAnimating];
    
    // Cancel all pending fetches and decompress image operations
    for (GMCDecompressImageOperation *decompressImageOperation in self.decompressImageOperations) {
        [decompressImageOperation cancel];
    }
    for (GMCMultiImageRenditionFetch *multiImageRenditionFetch in self.multiImageRenditionFetches) {
        [multiImageRenditionFetch cancel];
    }
    [self.decompressImageOperations removeAllObjects];
    [self.multiImageRenditionFetches removeAllObjects];
    
    // Resize image view
    CGSize fullImageSize = [self.multiImage largestRenditionSize];
    self.imageView.frame = CGRectMake(0, 0, fullImageSize.width, fullImageSize.height);
    
    // Resize scroll view and establish initial zoom and zoom limits
    self.scrollView.contentSize = fullImageSize;
    self.scrollView.minimumZoomScale = MIN(1, MIN(self.scrollView.bounds.size.width / fullImageSize.width, self.scrollView.bounds.size.height / fullImageSize.height));
    self.scrollView.zoomScale = self.scrollView.minimumZoomScale;
    
    self.initialZoomScale = self.scrollView.minimumZoomScale;
    self.otherZoomScale = 1;
    
    [self updateImage];
    [self centerImage];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    if (!CGSizeEqualToSize(self.previousBoundsSize, self.bounds.size)) {
        self.previousBoundsSize = self.bounds.size;
        
        CGSize fullImageSize = [self.multiImage largestRenditionSize];
        
        CGFloat previousMinimumZoomScale = self.scrollView.minimumZoomScale;
        self.scrollView.minimumZoomScale = MIN(1, MIN(self.scrollView.bounds.size.width / fullImageSize.width, self.scrollView.bounds.size.height / fullImageSize.height));
        if (self.scrollView.zoomScale == previousMinimumZoomScale) {
            self.scrollView.zoomScale = self.scrollView.minimumZoomScale;
        }
        self.scrollView.zoomScale = MAX(self.scrollView.minimumZoomScale, self.scrollView.zoomScale);
        
        self.initialZoomScale = self.scrollView.minimumZoomScale;
        
        [self updateImage];
        [self centerImage];
    }
}

- (void)centerImage {
    CGPoint center;
	center.x = roundf(self.scrollView.contentSize.width / 2 + MAX(0, (self.scrollView.bounds.size.width - self.scrollView.contentSize.width) / 2));
	center.y = roundf(self.scrollView.contentSize.height / 2 + MAX(0, (self.scrollView.bounds.size.height - self.scrollView.contentSize.height) / 2));
    self.imageView.center = center;
}

- (void)zoomToScale:(CGFloat)scale atLocation:(CGPoint)location {
	CGRect zoomRect;
    zoomRect.size.width  = self.scrollView.bounds.size.width / scale;
	zoomRect.size.height = self.scrollView.bounds.size.height / scale;
    zoomRect.origin.x = location.x - (zoomRect.size.width / 2.0);
    zoomRect.origin.y = location.y - (zoomRect.size.height / 2.0);
	[self.scrollView zoomToRect:zoomRect animated:YES];
}

- (void)updateImage {
    if (!self.multiImage) {
        return;
    }
    
    CGSize fullImageSize = [self.multiImage largestRenditionSize];
    CGSize desiredSize = CGSizeMake(fullImageSize.width * self.scrollView.zoomScale, fullImageSize.height * self.scrollView.zoomScale);
    
    GMCMultiImageRendition *rendition = [self.multiImage bestRenditionThatFits:desiredSize scale:self.scale contentMode:GMCMultiImageContentModeScaleAspectFit];
    if (![self.currentRendition isEqual:rendition]) {
        self.currentRendition = rendition;
        
        if (self.imageView.image == nil) {
            GMCMultiImageRendition *smallestRendition = [self.multiImage bestRenditionThatFits:self.placeholderSize scale:self.scale contentMode:GMCMultiImageContentModeScaleAspectFit];
            if (smallestRendition.isImageAvailable) {
                [self.loadingIndicatorView stopAnimating];
                
                // Show the smallest representation first, if available.
                // It's OK to do this on the main thread because the image is very small.
                self.imageView.image = smallestRendition.image;
            } else {
                [self.loadingIndicatorView startAnimating];
                
                // If the smallest representation is not available, go ahead and fetch it, then show it
                __block GMCMultiImageRenditionFetch *fetch = [smallestRendition fetchImageWithCompletionBlock:^(NSError *error) {
                    if ([self.multiImageRenditionFetches containsObject:fetch]) {
                        [self.multiImageRenditionFetches removeObject:fetch];
                    }
                    
                    if (!error) {
                        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
                            if ([self.currentRendition isEqual:rendition] && self.imageView.image == nil) {
                                GMCDecompressImageOperation *operation = [[GMCDecompressImageOperation alloc] init];
                                operation.image = smallestRendition.image;
                                
                                __weak GMCDecompressImageOperation *weakOperation = operation;
                                operation.completionBlock = ^{
                                    UIImage *image = weakOperation.image;
                                    if ([self.decompressImageOperations containsObject:weakOperation]) {
                                        [self.decompressImageOperations removeObject:weakOperation];
                                    }
                                    
                                    dispatch_async(dispatch_get_main_queue(), ^{
                                        if ([self.currentRendition isEqual:rendition] && self.imageView.image == nil) {
                                            [self.loadingIndicatorView stopAnimating];
                                            
                                            self.imageView.image = image;
                                        }
                                    });
                                };
                                
                                [self.decompressImageOperations addObject:operation];
                                [[NSOperationQueue decompressImageQueue] addOperation:operation];
                            }
                        });
                    }
                }];
                if (fetch) {
                    [self.multiImageRenditionFetches addObject:fetch];
                }
            }
        }
        
        if (!rendition.isImageAvailable) {
            GMCMultiImageRendition *bestAvailableRendition = [self.multiImage bestAvailableRenditionThatFits:desiredSize scale:self.scale contentMode:GMCMultiImageContentModeScaleAspectFit];
            if (bestAvailableRendition && ![bestAvailableRendition isEqual:rendition]) {
                // Fetch and set the best available representation, as long as the desired one hasn't been set yet.
                // Don't directly set the image, as that would cause the image to be loaded on the main thread.
                __block GMCMultiImageRenditionFetch *fetch = [bestAvailableRendition fetchImageWithCompletionBlock:^(NSError *error) {
                    if ([self.multiImageRenditionFetches containsObject:fetch]) {
                        [self.multiImageRenditionFetches removeObject:fetch];
                    }
                    
                    if (!error) {
                        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
                            if ([self.currentRendition isEqual:rendition]) {
                                GMCDecompressImageOperation *operation = [[GMCDecompressImageOperation alloc] init];
                                operation.image = bestAvailableRendition.image;
                                
                                __weak GMCDecompressImageOperation *weakOperation = operation;
                                operation.completionBlock = ^{
                                    UIImage *image = weakOperation.image;
                                    if ([self.decompressImageOperations containsObject:weakOperation]) {
                                        [self.decompressImageOperations removeObject:weakOperation];
                                    }
                                    
                                    dispatch_async(dispatch_get_main_queue(), ^{
                                        if ([self.currentRendition isEqual:rendition]) {
                                            [self.loadingIndicatorView stopAnimating];
                                            
                                            self.imageView.image = image;
                                        }
                                    });
                                };
                                
                                [self.decompressImageOperations addObject:operation];
                                [[NSOperationQueue decompressImageQueue] addOperation:operation];
                            }
                        });
                    }
                }];
                if (fetch) {
                    [self.multiImageRenditionFetches addObject:fetch];
                }
            }
        }
        
        // Fetch and set the desired image representation.
        // Don't directly set the image, as that would cause the image to be loaded on the main thread.
        __block GMCMultiImageRenditionFetch *fetch = [rendition fetchImageWithCompletionBlock:^(NSError *error) {
            if ([self.multiImageRenditionFetches containsObject:fetch]) {
                [self.multiImageRenditionFetches removeObject:fetch];
            }
            
            if (error) {
                if ([self.currentRendition isEqual:rendition]) {
                    [self.loadingIndicatorView stopAnimating];
                }
            } else {
                dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
                    if ([self.currentRendition isEqual:rendition]) {
                        GMCDecompressImageOperation *operation = [[GMCDecompressImageOperation alloc] init];
                        operation.image = rendition.image;
                        
                        __weak GMCDecompressImageOperation *weakOperation = operation;
                        operation.completionBlock = ^{
                            UIImage *image = weakOperation.image;
                            if ([self.decompressImageOperations containsObject:weakOperation]) {
                                [self.decompressImageOperations removeObject:weakOperation];
                            }
                            
                            dispatch_async(dispatch_get_main_queue(), ^{
                                if ([self.currentRendition isEqual:rendition]) {
                                    [self.loadingIndicatorView stopAnimating];
                                    
                                    self.imageView.image = image;
                                }
                            });
                        };
                        
                        [self.decompressImageOperations addObject:operation];
                        [[NSOperationQueue decompressImageQueue] addOperation:operation];
                    }
                });
            }
        }];
        if (fetch) {
            [self.multiImageRenditionFetches addObject:fetch];
        }
    }
}

#pragma mark - Gesture Recognizers

- (void)doubleTap:(UITapGestureRecognizer *)gestureRecognizer {
    CGPoint location = [gestureRecognizer locationInView:self.imageView];
    if (!fequal(self.scrollView.zoomScale, self.initialZoomScale)) {
        [self zoomToScale:self.initialZoomScale atLocation:location];
    } else if (!fequal(self.initialZoomScale, self.otherZoomScale)) {
        [self zoomToScale:self.otherZoomScale atLocation:location];
    }
}

#pragma mark - UIScrollViewDelegate

- (UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView {
    return self.imageView;
}

- (void)scrollViewDidEndZooming:(UIScrollView *)scrollView withView:(UIView *)view atScale:(float)scale {
    [self updateImage];
}

- (void)scrollViewDidZoom:(UIScrollView *)scrollView {
    [self centerImage];
}

@end

File: GMCMultiImage/GMCZoomingMultiImageView.swift
//
// Copyright (c) 2018 Hilton Campbell
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import UIKit

public class GMCZoomingMultiImageView: UIView, UIScrollViewDelegate {
    
    static let PlaceholderSizeDefault = CGSize(width: 55, height: 55)
    
    var placeholderSize = CGSize.zero
    var scale: CGFloat = 1.0
    
    private lazy var scrollView: UIScrollView = {
        let scrollView = UIScrollView(frame: self.bounds)
        #if !TARGET_OS_TV
        scrollView.scrollsToTop = false
        #endif
        scrollView.showsVerticalScrollIndicator = false
        scrollView.showsHorizontalScrollIndicator = false
        scrollView.delegate = self
        scrollView.maximumZoomScale = 1
        scrollView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        return scrollView
    }()
    
    private var doubleTapGestureRecognizer: UITapGestureRecognizer?
    private lazy var imageView = UIImageView()
    
    private lazy var loadingIndicatorView: UIActivityIndicatorView = {
        let loadingIndicatorView = UIActivityIndicatorView(activityIndicatorStyle: .whiteLarge)
        loadingIndicatorView.frame = self.bounds
        loadingIndicatorView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        return loadingIndicatorView
    }()
    
    private var previousBoundsSize = CGSize.zero
    private var currentRendition: GMCMultiImageRendition?
    private var multiImageRenditionFetches: [Operation] = []
    private var decompressImageOperations: [Operation] = []
    private var initialZoomScale: CGFloat = 0.0
    private var otherZoomScale: CGFloat = 0.0
    
    public var multiImage: GMCMultiImage? {
        didSet {
            currentRendition = nil
            imageView.image = nil
            loadingIndicatorView.stopAnimating()
            // Cancel all pending fetches and decompress image operations
            for decompressImageOperation in decompressImageOperations {
                decompressImageOperation.cancel()
            }
            for multiImageRenditionFetch in multiImageRenditionFetches {
                multiImageRenditionFetch.cancel()
            }
            decompressImageOperations.removeAll()
            multiImageRenditionFetches.removeAll()
            // Resize image view
            
            if let fullImageSize = multiImage?.largestRenditionSize {
                imageView.frame = CGRect(x: 0, y: 0, width: fullImageSize.width, height: fullImageSize.height)
            }
            
            updateZoomInitial(true)
            updateImage()
            centerImage()
        }
    }
    
    override public var contentMode: UIViewContentMode {
        didSet {
            updateZoomInitial(false)
        }
    }
    
    override public init(frame: CGRect) {
        super.init(frame: frame)
        
        contentMode = .scaleAspectFit
        
        addSubview(scrollView)
        scrollView.addSubview(imageView)
        
        addSubview(loadingIndicatorView)
        
        multiImageRenditionFetches = [Operation]()
        decompressImageOperations = [Operation]()
        
        let doubleTapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(doubleTap))
        doubleTapGestureRecognizer.numberOfTapsRequired = 2
        addGestureRecognizer(doubleTapGestureRecognizer)
        self.doubleTapGestureRecognizer = doubleTapGestureRecognizer
        
        placeholderSize = GMCZoomingMultiImageView.PlaceholderSizeDefault
        scale = UIScreen.main.scale
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override public func layoutSubviews() {
        super.layoutSubviews()
        
        guard !previousBoundsSize.equalTo(bounds.size) else { return }
        
        previousBoundsSize = bounds.size
        updateZoomInitial(false)
        updateImage()
        centerImage()
    }
    
    func updateZoomInitial(_ initial: Bool) {
        guard let fullImageSize = multiImage?.largestRenditionSize else { return }
        
        scrollView.contentSize = fullImageSize
        
        let contentFrame: CGRect = UIEdgeInsetsInsetRect(scrollView.bounds, scrollView.contentInset)
        let previousMinimumZoomScale: CGFloat = scrollView.minimumZoomScale
        
        switch contentMode {
        case .scaleAspectFill:
            scrollView.minimumZoomScale = max(contentFrame.size.width / fullImageSize.width, contentFrame.size.height / fullImageSize.height)
        default:
            scrollView.minimumZoomScale = min(1, min(contentFrame.size.width / fullImageSize.width, contentFrame.size.height / fullImageSize.height))
        }
        initialZoomScale = scrollView.minimumZoomScale
        
        if initial {
            scrollView.zoomScale = scrollView.minimumZoomScale
            otherZoomScale = 1
        } else {
            if scrollView.zoomScale == previousMinimumZoomScale {
                scrollView.zoomScale = scrollView.minimumZoomScale
            }
            scrollView.zoomScale = max(scrollView.minimumZoomScale, scrollView.zoomScale)
        }
    }
    
    func centerImage() {
        let contentFrame = UIEdgeInsetsInsetRect(scrollView.bounds, scrollView.contentInset)
        
        imageView.center = CGPoint(x: round(scrollView.contentSize.width / 2.0 + max(0.0, (contentFrame.size.width - scrollView.contentSize.width) / 2.0)),
                                    y: round(scrollView.contentSize.height / 2.0 + max(0.0, (contentFrame.size.height - scrollView.contentSize.height) / 2.0)))
    }
    
    func zoom(toScale scale: CGFloat, atLocation location: CGPoint) {
        let zoomWidth = scrollView.bounds.size.width / scale
        let zoomHeight = scrollView.bounds.size.height / scale
        
        let zoomRect = CGRect(x: location.x - (zoomWidth / 2.0),
                              y: location.y - (zoomHeight / 2.0),
                              width: zoomWidth,
                              height: zoomHeight)
        
        scrollView.zoom(to: zoomRect, animated: true)
    }
    
    func updateImage() {
        guard let multiImage = multiImage, let fullImageSize = multiImage.largestRenditionSize else { return }
        
        let desiredSize = CGSize(width: fullImageSize.width * scrollView.zoomScale, height: fullImageSize.height * scrollView.zoomScale)
        let rendition = multiImage.bestRenditionThatFits(size: desiredSize, scale: scale, contentMode: .scaleAspectFit)
        
        guard currentRendition != rendition else { return }
        
        currentRendition = rendition
        
        if imageView.image == nil {
            let smallestRendition = multiImage.bestRenditionThatFits(size: placeholderSize, scale: scale, contentMode: .scaleAspectFit)
            if smallestRendition?.isImageAvailable == true {
                // Show the smallest representation first, if available.

                DispatchQueue.main.async { [weak self] in
                    self?.loadingIndicatorView.stopAnimating()
                    self?.imageView.image = smallestRendition?.image
                }
            } else {
                DispatchQueue.main.async { [weak self] in
                    self?.loadingIndicatorView.startAnimating()
                }
                
                // If the smallest representation is not available, go ahead and fetch it, then show it
                smallestRendition?.fetchImage { [weak self] error, image  in
                    guard error == nil, self?.currentRendition == rendition else { return }
                    
                    let operation = GMCDecompressImageOperation()
                    operation.image = smallestRendition?.image
                    operation.completionBlock = { [weak self] () in
                        DispatchQueue.main.async { [weak self] in
                            guard self?.currentRendition == rendition && self?.imageView.image == nil else { return }
                        
                            self?.loadingIndicatorView.stopAnimating()
                            self?.imageView.image = operation.image
                        }
                    }
                    self?.decompressImageOperations.append(operation)
                    OperationQueue.decompressImageQueue.addOperation(operation)
                }
                
                if rendition?.isImageAvailable == false, let bestAvailableRendition = multiImage.bestAvailableRenditionThatFits(size: desiredSize, scale: scale, contentMode: .scaleAspectFit), bestAvailableRendition != rendition {
                    bestAvailableRendition.fetchImage { [weak self] error, image in
                        guard self?.currentRendition == rendition else { return }
                        guard error == nil else {
                            DispatchQueue.main.async { [weak self] in
                                self?.loadingIndicatorView.stopAnimating()
                            }
                            return
                        }
                        
                        let operation = GMCDecompressImageOperation()
                        operation.image = bestAvailableRendition.image
                        operation.completionBlock = { [weak self] () in
                            guard self?.currentRendition == rendition else { return }
                            
                            DispatchQueue.main.async { [weak self] in
                                self?.loadingIndicatorView.stopAnimating()
                                self?.imageView.image = operation.image
                            }
                        }
                        self?.decompressImageOperations.append(operation)
                        OperationQueue.decompressImageQueue.addOperation(operation)
                    }
                }
            }
        } else {
            rendition?.fetchImage { [weak self] error, image in
                guard self?.currentRendition == rendition else { return }
                guard error == nil else {
                    DispatchQueue.main.async { [weak self] in
                        self?.loadingIndicatorView.stopAnimating()
                    }
                    return
                }
                
                let operation = GMCDecompressImageOperation()
                operation.image = rendition?.image
                operation.completionBlock = { [weak self] () in
                    guard self?.currentRendition == rendition else { return }
                    
                    DispatchQueue.main.async { [weak self] in
                        self?.loadingIndicatorView.stopAnimating()
                        self?.imageView.image = operation.image
                    }
                }
                self?.decompressImageOperations.append(operation)
                OperationQueue.decompressImageQueue.addOperation(operation)
            }
        }
    }
    
    // MARK: - Gesture Recognizers
    @objc func doubleTap(_ gestureRecognizer: UITapGestureRecognizer) {
        let location = gestureRecognizer.location(in: imageView)
        if scrollView.zoomScale != initialZoomScale {
            zoom(toScale: initialZoomScale, atLocation: location)
        } else if initialZoomScale != otherZoomScale {
            zoom(toScale: otherZoomScale, atLocation: location)
        }
    }
    
    // MARK: - UIScrollViewDelegate
    public func viewForZooming(in scrollView: UIScrollView) -> UIView? {
        return imageView
    }
    
    public func scrollViewDidEndZooming(_ scrollView: UIScrollView, with view: UIView?, atScale scale: CGFloat) {
        updateImage()
    }
    
    public func scrollViewDidZoom(_ scrollView: UIScrollView) {
        centerImage()
    }
    
}
